Docs about different configs [WIP] (#2442)
conan-io/docs: tutorial/consuming_packages/getting_started/different_configurations.rst

Building for multiple configurations: Release, Debug, Static and Shared
=======================================================================

.. important::

    In this example, we will retrieve the CMake Conan package from a Conan repository with

>     packages compatible for Conan 2.0. To run this example succesfully you should add this

Reviewer: ```suggestion
    packages compatible with Conan 2.0. To run this example successfully you should add this
```

    remote to your Conan configuration (if did not already do it) doing:
    ``conan remote add conanv2 https://conanv2beta.jfrog.io/artifactory/api/conan/conan --index 0``

Please, first clone the sources to recreate this project. You can find them in the
`examples2.0 repository <https://github.com/conan-io/examples2>`_ in GitHub:

.. code-block:: bash

    $ git clone https://github.com/conan-io/examples2.git

>     $ cd tutorial/consuming_packages/getting_started/different_configurations

Reviewer: ```suggestion
    $ cd examples2/tutorial/consuming_packages/getting_started/different_configurations
```

So far, we built a simple CMake project that depended on the **zlib** library and learned

> about ``tool_requires``, a special type or requirements for build tools like CMake. In

Reviewer: ```suggestion
about ``tool_requires``, a special type or requirements for build-tools like CMake. In
```

> both cases we did not specify anywhere that we wanted to build the application in

Reviewer: ```suggestion
both cases, we did not specify anywhere that we wanted to build the application in
```

*Release* or *Debug* mode, or if we wanted to link against *static* or *shared* libraries.
That is because Conan, if not instructed otherwise, will use a default configuration
declared in the 'default profile'. This default profile was created in the first example
when we run the ``conan profile detect`` command. Conan stores this file in the **/profiles**
folder, located in the Conan user home. You can check the contents of your default
profile:

> Run the ``conan config`` command and get the location of the Conan user home, then show

Reviewer: ```suggestion
Run the ``conan config home`` command and get the location of the Conan user home, then show
```

the contents of the default profile:

.. code-block:: bash

    $ conan config home
    Current Conan home: /Users/tutorial_user/.conan2

    # output the file contents
    $ cat /Users/tutorial_user/.conan2/profiles/default
    [settings]
    os=Macos
    arch=x86_64
    compiler=apple-clang
    compiler.version=13.0
    compiler.libcxx=libc++
    compiler.cppstd=gnu98
    build_type=Release
    [options]
    [tool_requires]
    [env]

> As you can see, the profile have different sections. The ``[settings]`` section is the one

Reviewer: ```suggestion
As you can see, the profile has different sections. The ``[settings]`` section is the one
```

> that has information about things like the operating system, architecture, compiler and

Reviewer: ```suggestion
that has information about things like the operating system, architecture, compiler, and
```

build configuration. When you call a Conan command setting the ``--profile`` argument,
Conan will take all the information from the profile and apply it to the packages you want

> to build or instal. If you don't specify that argument it's equivalent to call it with

Reviewer: ```suggestion
to build or install. If you don't specify that argument it's equivalent to call it with
```

> ``--profile=default``. These two commands will behave exactly the same:

Reviewer: ```suggestion
``--profile=default``. These two commands will behave the same:
```

.. code-block:: bash

    $ conan install . --build=missing
    $ conan install . --build=missing --profile=default

> You can store different profiles and use them to build for differente setings. For example,

Reviewer: ```suggestion
You can store different profiles and use them to build for different settings. For example,
```

example, you can build the project from the previous examples in *Debug* configuration
instead of *Release*.

Before building, please check that we modified the source code from the previous example to show the build configuration the
sources were built with:

.. code-block:: cpp
    :emphasize-lines: 6-10

    #include <stdlib.h>
    ...

    int main(void) {
        ...
        #ifdef NDEBUG
        printf("Release configuration!\n");
        #else
        printf("Debug configuration!\n");
        #endif

        return EXIT_SUCCESS;
    }

Now let's build our project for *Debug* configuration:

.. code-block:: bash
    :caption: Windows

    $ conan install . --output-folder=build --build=missing --settings=build_type=Debug

.. code-block:: bash
    :caption: Linux, macOS
    
    $ conan install . --output-folder cmake-build-release --build=missing --settings=build_type=Debug

This ``conan install`` command will check if we already installed the required libraries
(Zlib) in Debug configuration and install them otherwise. It will also set the build
configuration in the ``conan_toolchain.cmake`` toolchain that the CMakeToolchain generator

> creates so that when we build the application it's build in *Debug* configuration. Now

Reviewer: ```suggestion
creates so that when we build the application it's built in *Debug* configuration. Now
```

build your project as you did in the previous examples and check in the output how it was
built in *Debug* configuration:

.. code-block:: bash
    :caption: Windows
    :emphasize-lines: 8

    # assuming Visual Studio 15 2017 is your VS version and that it matches your default profile

>     $ cmake .. -G "Visual Studio 15 2017" -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake

Reviewer: ```suggestion
    $ cd build
    $ cmake .. -G "Visual Studio 15 2017" -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake
```

    $ cmake --build . --config Debug
    $ Debug\compressor.exe
    Uncompressed size is: 233
    Compressed size is: 147
    ZLIB VERSION: 1.2.11
    Debug configuration!

.. code-block:: bash
    :caption: Linux, macOS
    :emphasize-lines: 7
    

>     $ cmake .. -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake

Reviewer: ```suggestion
    $ cd cmake-build-release
    $ cmake .. -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake
```

    $ cmake --build .
    $ ./compressor
    Uncompressed size is: 233
    Compressed size is: 147
    ZLIB VERSION: 1.2.11
    Debug configuration!

.. _different_configurations_modify_options:

Modifying options: linking the application dependencies as shared libraries
---------------------------------------------------------------------------

So far, we have been linking *Zlib* statically in our application. That's because in the
Zlib's Conan package there's an attribute set to build in that mode by default. We can
change from **static** to **shared** linking by setting the ``shared`` option to ``True``
using the ``--options`` argument. To do so, please run:

.. code-block:: bash
    :caption: Windows

    $ conan install . --output-folder=build --build=missing --options=zlib/1.2.11:shared=True

.. code-block:: bash
    :caption: Linux, macOS
    
    $ conan install . --output-folder cmake-build-release --build=missing --options=zlib/1.2.11:shared=True

Doing this, Conan will install the *Zlib* shared libraries, generate the files to build with

> them and also the necessary files to locate those dynamic libraries when running the

Reviewer: ```suggestion
them and, also the necessary files to locate those dynamic libraries when running the
```

    $ cmake .. -G "Visual Studio 15 2017" -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake
    $ cmake --build . --config Release
    ...
    [100%] Built target compressor

.. code-block:: bash
    :caption: Linux, Macos
    
    $ cd cmake-build-release
    $ cmake .. -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake
    $ cmake --build .
    ...
    [100%] Built target compressor

Now, if you try to run the compiled executable you will see an error because the
executable can't find the shared libraries for *Zlib* that we just installed.

.. code-block:: bash
    :caption: Windows

    $ Release\compressor.exe
    (on a pop-up window) The code execution cannot proceed because zlib1.dll was not found. Reinstalling the program may fix this problem.

.. code-block:: bash
    :caption: Linux, Macos
    
    $ ./compressor
    ./compressor: error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory

This is because shared libraries (*.dll* in windows, *.dylib* in OSX and *.so* in Linux),
are loaded at runtime. That means that the application executable needs to know where are

> the required shared libraries when it runs. On Windows, the dynamic linker, will search in

Reviewer: ```suggestion
the required shared libraries when it runs. On Windows, the dynamic linker will search in
```

the same directory then in the *PATH* directories. On OSX, it will search in the
directories declared in *DYLD_LIBRARY_PATH* as on Linux will use the *LD_LIBRARY_PATH*.

> Conan provides a mechanism to define those variables and make possible for executables to

Reviewer: ```suggestion
Conan provides a mechanism to define those variables and make it possible, for executables, to
```

``conanrun.sh/bat`` script a deactivation script called ``deactivate_conanrun.sh/bat`` is
created to restore the environment. Source or run it to do so:

.. code-block:: bash
    :caption: Windows

    $ deactivate_conanrun.bat

.. code-block:: bash
    :caption: Linux, macOS
    
    $ source deactivate_conanrun.sh

Difference between settings and options
=======================================

You may have noticed that for changing between *Debug* and *Release* configuration we
used a Conan **setting**, but when we set *shared* mode for our executable we used a
Conan **option**. Please, note the difference between **settings** and **options**:

* **settings** are typically a project-wide configuration defined by the client machine.
  Things like the operating system, compiler or build configuration that will be common to
  several Conan packages and would not make sense to define one default value for only one
  of them. For example, it doesn’t make sense for a Conan package to declare “Visual
  Studio” as a default compiler because that is something defined by the end consumer, and
  unlikely to make sense if they are working in Linux.

> * **options** are intended for package specific configuration that can be set to a default

Reviewer: ```suggestion
* **options** are intended for package-specific configuration that can be set to a default
```

  value in the recipe. For example, one package can define that its default linkage is
  static, and this is the linkage that should be used if consumers don’t specify
  otherwise.

Read more
---------

- Installing configurations with conan config install
- VS Multi-config
- Example about how settings and options influence the package id

> - Cross compiling using --profile:build and --profile:host

Reviewer: ```suggestion
- Cross-compiling using --profile:build and --profile:host
```
